File: Algorithms/JavaScript/all-anagrams.js
/**
 * Given a single input string, write a function that produces all possible anagrams
 * of a string and outputs them as an array. At first, don't worry about
 * repeated strings.  What time complexity is your solution?
 *
 * Extra credit: Deduplicate your return array without using uniq().
 */

/**
  * example usage:
  * var anagrams = allAnagrams('abc');
  * console.log(anagrams); // [ 'abc', 'acb', 'bac', 'bca', 'cab', 'cba' ]
  */

var allAnagrams = function(string) {

  if (typeof(string) !== 'string') {return;}

  if (string.length === 1) {
    return [string];
  } else {
    var anagrams = [],
        dict = {},
        current,
        others;

    for (var i = 0; i < string.length; i++){
      current = string[i];
      others = allAnagrams(string.substring(0, i) + string.substring(i + 1))
      for (var j = 0; j < others.length; j++){
        var concatenated = current.concat(others[j]);
        if (!dict.hasOwnProperty(concatenated)){
          dict[concatenated] = true;
          anagrams.push(concatenated);
        }
      }    
    }

    return anagrams;
  }

};

File: Algorithms/Ruby/all-anagrams.rb
# Given a single input string, write a function that produces all possible anagrams
# of a string and outputs them as an array. At first, don't worry about
# repeated strings.  What time complexity is your solution?
#
# Extra credit: Deduplicate your return array without using uniq().
#

def all_anagrams(string)
  return false unless string.is_a?(String)
  return [string] if string.length == 1
  anagrams, dict = [], {}

  (0...string.length).each do |i|
    current = string[i]
    others = all_anagrams(string[0...i] + string[i+1..-1])
    (0...others.length).each do |j|
      concatenated = current + others[j]
      unless dict[concatenated]
        dict[concatenated] = true
        anagrams.push(concatenated)
      end
    end
  end

  anagrams
end

all_anagrams("abc")
